src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs

```diff
@@ -428,33 +427,27 @@ public void Dispose()
             internal void FunctionStarted(FunctionStartedEvent startedEvent)
             {
                 _totalExecutionCount++;
+                Interlocked.Increment(ref _activeFunctionCount);
 
                 var monitoringEvent = new FunctionMetrics(startedEvent.FunctionMetadata.Name, ExecutionStage.Started, 0);
                 _functionMetricsQueue.Enqueue(monitoringEvent);
 
-                var key = GetDictionaryKey(startedEvent.FunctionMetadata.Name, startedEvent.InvocationId);
-                var triggerType = startedEvent.FunctionMetadata.Trigger?.Type;
-                var value = new RunningFunctionInfo(startedEvent.FunctionMetadata.Name, startedEvent.InvocationId, startedEvent.Timestamp, startedEvent.Success, triggerType);
-                _runningFunctions.AddOrUpdate(key, value, (k, v) => value);
+                lock (_runningFunctionsSyncLock)
+                {
+                    _runningFunctions.Add(startedEvent);
+                }
             }
 
             internal void FunctionCompleted(FunctionStartedEvent startedEvent)
             {
-                var functionStage = (startedEvent.Success == false) ? ExecutionStage.Failed : ExecutionStage.Succeeded;
+                Interlocked.Decrement(ref _activeFunctionCount);
+
+                var functionStage = !startedEvent.Success ? ExecutionStage.Failed : ExecutionStage.Succeeded;
                 long executionTimeInMS = (long)startedEvent.Duration.TotalMilliseconds;
                 var monitoringEvent = new FunctionMetrics(startedEvent.FunctionMetadata.Name, functionStage, executionTimeInMS);
                 _functionMetricsQueue.Enqueue(monitoringEvent);
 
-                var key = GetDictionaryKey(startedEvent.FunctionMetadata.Name, startedEvent.InvocationId);
-                if (_runningFunctions.TryRemove(key, out RunningFunctionInfo functionInfo))
```

Comment:
We no longer remove from the collection on completion, reducing lock contention. We just raise the completed event. The background timer that is raising in progress events will skip/purge any completed functions.